fix: Adds support for defining Interface entities resolvers

moonflare opened the pull request.

moonflare committed:
feat: adds support for Interface Entities

Context: In Federation v2.3 there was added support for interfaces to be
Entities. There are two ways to consume this new feature.

1. Extend an Interface Entity with the `@interfaceObject` directive,
   when the Entity is defined by a different subgraph.

2. Implement the Interface Entity with its implementing Type Entitites.

There was previos work into getting the first part covered in this gem,
but we were still lacking support for the 2nd part. This work address
that.

The main issue here is the fact that an `Union` in `GraphQL` can't be an
`Interface` according to the [spec](https://spec.graphql.org/October2021/#sec-Unions.Type-Validation), but at the same time, according to the Apollo Federation [spec](https://www.apollographql.com/docs/federation/federated-types/interfaces), an interface can be an Entity, and an Entity is an Union. Therefore, we have to extend the validation (`assert_valid_union_member`) for the Entity union to allow `Interfaces` (more exactly `Modules`) as possible types.

moonflare committed:
fix: adds `underscore_reference_keys` support to interfaces

In a previous Pull Request (https://github.com/Gusto/apollo-federation-ruby/pull/248) there was added support for underscoring the keys and while this covered the `Object` class, it didn't also cover the `Interface` module. This commit adds the same `underscore_reference_keys` method to the `Interface` module.

moonflare committed:
docs: adds docs for Interface Entities

nogates commented on lib/apollo-federation/entity.rb:
```diff
@@ -9,5 +9,25 @@ class Entity < GraphQL::Schema::Union
+    # but at the same time, according to the Federation spec, an interface can be an Entity,
+    # and an Entity is an union. Therefore, we have to extend this validation to allow interfaces
+    # (more precisely Modules) as possible types.
+    def self.assert_valid_union_member(type_defn)
```
I think it would be good if we can monkey patch as little as possible. 
Also, double check that the module is indeed a GraphQL::Interface? 
Suggestion

```
def self.assert_valid_union_member(type_defn)
      if type_defn.is_a?(Module) &&
          type_defn.included_modules.include?(GraphQL::Schema::Interface)
        # This is the only change we made to the original method
        # to allow interfaces as possible types
      else
        super(type_defn)
      end
    end

```

moonflare replied:
Thanks for the suggestion. I've actually changed it a bit to check straight for `ApolloFederation::Interface`, which makes most sense in this context.

nogates committed:
fix: ensure all interface members are also entities

Every entity that implements an interface entity must include all `@keys` from the interface definition.
Therefore, we add this extra validation to ensure that.
